nixos/prosody: add MUC extraConfig + fixes

mmilata opened the pull request:
Thanks for the work in #86067 @NinjaTrappeur, even though it probably wasn't the goal it makes the Jitsi Meet module compose better with user-defined Prosody configuration! I'd like to propose for a few modifications in this PR:

* move the main `extraConfig` before the component definitions in the config file, so that it actually [applies globally](https://prosody.im/doc/configure#overview) instead of to the http_upload component
* add `extraConfig` option to the MUC submodule so that it's possible to provide additional options (like `storage` or `admins`)
* fix `nixos/tests/xmpp/prosod-mysql.nix`

cc @flokli 

###### Motivation for this change

Mostly #82920.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

NinjaTrappeur commented:
Thanks! Sounds good!

Aha, nice catch regarding the mysql test.

@GrahamcOfBorg test prosody-mysql

picnoir reviewed (approved):
LGTM!

mmilata commented:
Looks like `xmpp/ejabberd.nix` is also broken though I'm not familiar enough with it to fix it:/

mmilata commented:
Btw the jitsi module now only uses global `extraConfig` to add [external component](https://prosody.im/doc/components#adding_an_external_component):
```
        Component "focus.${cfg.hostName}"
          component_secret = os.getenv("JICOFO_COMPONENT_SECRET")
```

I'm wondering whether it would make sense to extend the Prosody module with options for external components. The `os.getenv` probably makes this needlessly complicated so I guess just adding this snippet to `extraConfig` with `mkAfter` minimizes the chance of bad ordering of `extraConfig`s from multiple sources.

NinjaTrappeur commented:
Indeed.

This test, together with the two prosody ones were broken for a long time actually. The `send-xmpp.py` script was always exiting with a 0 code, regardless of the script outcome. #86067 is trying its best to fix that :/

After a quick scan, it looks like the ebjabberd test is lacking the http_upload support, which we now aim to ship by default, we'll definitely need to setup this module for this test to pass.

I added this to my todo list.

This matter is however out of scope for this PR, let's merge this ;)

flokli commented on nixos/modules/services/networking/prosody.nix:
```diff
@@ -792,6 +797,8 @@ in
 
       https_ports = ${toLua cfg.httpsPorts}
 
+      ${ cfg.extraConfig }
```
Can you explain why this had to be moved before the muc definition?

NinjaTrappeur replied:
Not OP, but this makes impossible for the generic `extraConfig` to be interpreted as part of the MUC component config. It does make sense.

mmilata replied:
From [the manual](https://prosody.im/doc/configure#overview):
> The configuration is divided into two parts. The first part is known as the "global" section. All settings here apply to the whole server, and are the default for all virtual hosts.
> The second half of the file is a series of VirtualHost and Component definitions. Settings under each VirtualHost or Component line apply only to that host.

When `cfg.extraConfig` is rendered after the muc definition then it only applies to that muc and not globally, which is not how it behaved before and probably not what you want.

flokli replied:
Thanks for the explanation, and adding it to the commit message as well. This is much better :+1:

flokli commented on nixos/tests/xmpp/prosody-mysql.nix:
```diff
@@ -6,6 +6,11 @@ import ../make-test-python.nix {
       environment.systemPackages = [
         (pkgs.callPackage ./xmpp-sendmessage.nix { connectTo = nodes.server.config.networking.primaryIPAddress; })
       ];
+      networking.extraHosts = ''
+        ${nodes.server.config.networking.primaryIPAddress} example.com
+        ${nodes.server.config.networking.primaryIPAddress} conference.example.com
+        ${nodes.server.config.networking.primaryIPAddress} uploads.example.com
+      '';
```
The test fixes should be a separate commit

mmilata replied:
Done, PTAL.

mmilata committed:
nixos/prosody: add MUC extraConfig, fix extraConfig order

Add extraConfig option for the muc submodule.

Also move the global extraConfig before all components and
virtualhosts, because the manual states:

    The configuration is divided into two parts. The first part is known as
    the "global" section. All settings here apply to the whole server, and
    are the default for all virtual hosts.

    The second half of the file is a series of VirtualHost and Component
    definitions. Settings under each VirtualHost or Component line apply
    only to that host.

Before, if at least one muc was defined, or uploadHttp enabled, the
global extraConfig would end up after "muc" or "http_upload" component
making it apply to that component only and not globally.

mmilata committed:
nixosTests.prosodyMysql: fix

Since 8aea5288 xmpp-sendmessage.nix tests MUC and HTTP upload,
change the test to reflect this.
